File: Baekjoon/Data_Structure/Union-Find/20040_사이클_게임.py
# https://www.acmicpc.net/problem/20040

import sys
from collections import defaultdict

sys.setrecursionlimit(10**9)

def find(x: int) -> int:
    if tree[x] == -1:
        return x
    tree[x] = find(tree[x])
    return tree[x]

def union(x: int, y: int) -> None:
    x = find(x)
    y = find(y)
    if x != y:
        tree[x] = y

n, m = map(int, sys.stdin.readline().split())
tree = [-1] * n
flag = False

for i in range(m):
    a, b = map(int, sys.stdin.readline().split())

    if find(a) == find(b):
        print(i + 1)
        flag = True
        break
    else:
        union(a, b)

if not flag:
    print(0)

File: Baekjoon/Data_Structure/Disjoint_Sets/20040_사이클_게임.cpp
// https://www.acmicpc.net/problem/20040
#define FAST_IO ios_base::sync_with_stdio(0), cin.tie(0), cout.tie(0)

#include <iostream>
#include <algorithm>

using namespace std;

int tree[1000001] = { 0 };

int Find(int a) {
    if (tree[a] == -1) return a;
    tree[a] = Find(tree[a]);
    return tree[a];
}

void Union(int a, int b) { tree[a] = b; }

int main() {
    int a, b, i, n, m;
    FAST_IO;

    cin >> n >> m;
    fill_n(tree, n, -1);

    for (i = 0; i < m; i++) {
        cin >> a >> b;
        a = Find(a); b = Find(b);
        if (a == b) {
            cout << i + 1;
            return 0;
        }
        Union(a, b);
    }

    cout << 0;
    return 0;
}